Adding IES profile support to spot lights.

richardassar opened the pull request:
Implements the feature requested in https://github.com/mrdoob/three.js/issues/9624

- For demonstration/testing purposes I've just modified `examples/webgl_lights_spotlight`, I'll pull this out into its own example before we merge this PR.
- There are other uses of `getSpotLightInfo` that have still to be brought in line with these changes.
- Currently using the `IESLoader` that I updated, tidied and integrated into [gkjohnson/three-gpu-pathtracer](https://github.com/gkjohnson/three-gpu-pathtracer) as part of the spot light IES support there.
- Using gkjohnson's gist for the profile data itself, feel free to create some unique to three.js
- The profiles can (rarely) be two dimensional but this is ignored as it's a very rare case, I can add support for this if necessary.

Here's how it looks:

![Screenshot from 2022-07-30 16-34-04](https://user-images.githubusercontent.com/868396/181925628-42e7dd38-391d-4c45-84fc-56e386b01086.png)

![Screenshot from 2022-07-30 16-33-09](https://user-images.githubusercontent.com/868396/181925641-ab4a66c9-3e75-4c38-be76-4a76d97571a5.png)

Please review and I'll respond with changes accordingly.

richardassar committed:
Adding IES profile support to spot lights.

LeviPesin commented:
I don't think this approach ("integrating" IES profiles into core) is good. I think what was meant by @bhouston is rather loading IES profiles by parsing them and then creating lights using three.js' lights' own constructor parameters (i.e. intensity/distance/so on).
An alternative approach could be creating a `IESLight` class and making that IESLoader will produce such lights.

gkjohnson commented on src/renderers/shaders/ShaderChunk/lights_pars_begin.glsl.js:
```diff
@@ -144,9 +152,11 @@ float getSpotAttenuation( const in float coneCosine, const in float penumbraCosi
 	};
 
 	uniform SpotLight spotLights[ NUM_SPOT_LIGHTS ];
+	uniform sampler2D iesProfiles[ NUM_SPOT_LIGHTS ];
```
My only concern here is that with many spotlights you could quickly eat away at the max number of texture samplers even if no ies profiles are used on the spotlights. The texture array approach we used in gpu-pathtracer is the really generalized approach to this issue but I'm wondering if a simpler solution is to provide a minimal set of ies textures to the shader and use an index in the SpotLight struct? No ies profile would be indicated by - 1.

I'm not sure if this would work in WebGL 1, though, since I don't think it supports dynamically accessed arrays...

```glsl
	struct SpotLight {
		// ...
		float penumbraCos;
		int iesProfile;
	};
	uniform SpotLight spotLights[ NUM_SPOT_LIGHTS ];
	uniform sampler2D iesProfiles[ NUM_IES_PROFILES ];
```

richardassar replied:
I tried this. An issue I had is that I could not use iesProfiles[ spotLight.iesProfile ] as sampler arrays must be looked up with constant indices only, fine for the unrolled loop case.

richardassar replied:
Packing profiles into a single texture is a good approach, and if we stick with 1D profiles can just use `iesProfile` as one of the dimensions. Happy to go with this if you agree.

richardassar replied:
So we would have one main `sampler2D iesProfiles;` and just dynamically update that as new profiles are requested.

gkjohnson replied:
A concatenated texture would sound good to me and work in WebGL1 for 1D IES profiles as long we make sure that it doesn't get marked as needing reupload every render.

I'd wait for @mrdoob and @Mugen87 thoughts before updating the implementation, though.

richardassar replied:
Indeed, a `needsUpdate` only when the profile set is changed was what I was thinking.

richardassar replied:
Showing multiple 1D IES profiles that are combined into a single texture.

Now there's a single `uniform sampler2D iesProfiles;` and each light indexes into single lines of this.

![Screenshot from 2022-09-06 20-46-28](https://user-images.githubusercontent.com/868396/188725215-acbebd82-0015-4814-8c96-00bd1161d1d4.png)

Just needs a bit of tidy and then I'll submit for further review.

richardassar replied:
Slightly better screenshot of the above, easier to see the contribution of each light.

![Screenshot from 2022-09-07 08-45-57](https://user-images.githubusercontent.com/868396/188820101-478af3b0-ce3e-4ac4-976b-ff5501754b04.png)

gkjohnson commented:
> I think what was meant by @bhouston is rather loading IES profiles by parsing them and then creating lights using three.js' lights' own constructor parameters (i.e. intensity/distance/so on).

IES profiles cannot be replicated by just adjusting the existing constructor parameters alone for spot lights

LeviPesin commented:
Then I think the second suggestion is better. And I think IESLight and IESLoader would better go in examples rather than core...

gkjohnson commented:
> And I think IESLight and IESLoader would better go in examples rather than core...

If you look at the code it requires core shader changes...

richardassar commented:
> I don't think this approach ("integrating" IES profiles into core) is good... ~I think what was meant by @bhouston is rather loading IES profiles by parsing them and then creating lights using three.js' lights' own constructor parameters (i.e. intensity/distance/so on).~ An alternative approach could be creating a `IESLight` class which lights will IESLoader return.

I agree that if this is also required on point lights (although the distinction is artificial between point and spot lights with IES profiles) then a common base class `IESLight` which `IESPointLight` and `IESSpotLight` derive from seems like an acceptable solution.

UnrealEngine does allow one to set profiles on most light types, but again there's little difference except in the case of their "Rect Light" which just masks the profile leaving only one hemisphere.

The optional extra parameter to `SpotLight` seems unintrusive, but if there is consensus to refactor this as described above I'm happy to do so.

Mugen87 commented on src/loaders/IESLoader.js:
```diff
@@ -0,0 +1,325 @@
+
+}
+
+class IESLoader extends Loader {
```
Since this loader is very specific, I would move the module to `examples/jsm/loaders`.

richardassar replied:
I've made this change in https://github.com/mrdoob/three.js/pull/24409/commits/3aa0804b1690f90ee44b80dad0d4360ae809821f and https://github.com/mrdoob/three.js/pull/24409/commits/33609aa69d3fb1b44c65862bf9f96fa249f45f43.

Adding the change suggested in https://github.com/mrdoob/three.js/pull/24409#pullrequestreview-1056439383 and will ping for another review round.

richardassar committed:
Moving src/loaders/IESLoader.js to examples/jsm/loaders/IESLoader.js

richardassar committed:
Moving IESLoader to jsm/loaders

richardassar committed:
Added IESTexture.

- Takes one or an array of IES profile URLs
- Populates each row of the texture with each loaded IES profile, as rendered by IESLoader
- Increases the texture size by powers of two to amortize the cost of the copy

richardassar commented:
I've added `IESTexture` which accepts a single or array of IES profile URLs as an input to the constructor or the `add` method. Profiles can be added at creation time or later and `needsUpdate` will be flagged when appropriate.

As before the example remains in `webgl_lights_spotlight.html` (lights / spotlight). I will move this to it's own example and restore the original before we merge.

richardassar committed:
Optimised the Float32Array copies with .set and .subarray, and transposing the IESLoader output

richardassar commented:
@sunag pulled some of these changes into https://github.com/mrdoob/three.js/pull/25238

Ping @gkjohnson 

Would it be possible to get a review on this? 

I'd like to get #9624 closed, thanks.

gkjohnson reviewed (commented):
> Ping @gkjohnson
>
> Would it be possible to get a review on this?

Thanks for the reminder - I'm struggling to figure out how to cleanly add this to the existing three.js shader using existing patterns. Refreshing my mind on https://github.com/mrdoob/three.js/pull/24409#discussion_r933860457 - if there were a method for accessing arrays in `#define` conditions we could do something like this:

```glsl
uniform SpotLight spotLights[ NUM_SPOT_LIGHTS ];
uniform sampler2D iesProfiles[ NUM_IES_PROFILES ];

// ...

#unroll_loop_start
for ( ... ) {

    // ...
    #if SPOT_IES_PROFILE_INDEX[ UNROLLED_LOOP_INDEX ] != - 1

        float spotAttenuation = getPhotometricAttenuation( iesProfiles[ SPOT_IES_PROFILE_INDEX[ UNROLLED_LOOP_INDEX ] ], angleCos );

    #endif

}
#unroll_loop_end
```

This way the texture array is accessed with a constant index. @Mugen87 I seem to recall this being possible before. Has the capability been removed? Or am I misremembering? This could help simplify some of the spotlight map code quite a bit, as well.

richardassar commented on examples/webgl_lights_spotlight.html:
```diff
@@ -64,12 +66,29 @@
+				// iesProfilesTexture.add( 'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/06b4cfdc8805709e767b5e2e904be8ad.ies' );
+				iesProfilesTexture.add( 'https://raw.githubusercontent.com/gkjohnson/3d-demo-data/main/ies/02a7562c650498ebb301153dbbf59207.ies' );
+
+				scene.iesProfiles = iesProfilesTexture;
```
Thanks very much for that. 

Apologies, I had lost track of our earlier discussion, you also had asked @Mugen87 and @mrdoob in https://github.com/mrdoob/three.js/pull/24409#discussion_r933878109 for some feedback.

I think the constant index via `#define`s is a great workaround to avoid compatibility issues.

The proposal of

```js
spotLight.iesProfileMap = await new IESLoader().loadAsync( './path/to/profile.ies' );
```

seems good, much simpler.

Under the hood we can map IES profile paths to the profile index, which for 1D IES profiles is just an index into the horizontal or vertical line of the profiles texture. This also double as a caching mechanism to avoid ever loading and parsing IES profiles multiple times.

I'll do another pass over the code to implement the above and await feedback on the [spotlight logic for handling maps and shadows](https://github.com/mrdoob/three.js/blob/4c74ddc15b6a9eae758c0bc9c55c59fcf71d6464/src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js#L98-L110) issue.

gkjohnson replied:
It does seem that using a `#define` index to access a constant integer array works, though (but only in WebGL2...):

```glsl
uniform sampler2D textures[ 3 ];
const int indices[ 3 ] = int[]( 2, 1, 0 );

void main() {

  gl_FragColor = texture( textures[ indices[ 0 ] ], vec2( 0.5 ) );
    
}
```

https://jsfiddle.net/694jx3ny/1/

This still wouldn't let you only conditionally include the IES code per spotlight, though, since you can't access the constant array in a `#define` condition to check if it's - 1.

richardassar replied:
How about this?

```glsl

  #define ACCUM0 gl_FragColor += texture( textures[ indices0 ], vec2( 0.5 ) );
  #define ACCUM1 gl_FragColor += texture( textures[ indices1 ], vec2( 0.5 ) );
  #define ACCUM2 gl_FragColor += texture( textures[ indices2 ], vec2( 0.5 ) );
  
  #define LOOP0 ACCUM2  
  #define LOOP1 ACCUM1 LOOP0
  #define LOOP2 ACCUM0 LOOP1
  
  uniform sampler2D textures[3];
		
  #define indices0 2  
  #define indices1 1
  #define indices2 0

  void main() {
        
    gl_FragColor = vec4(0);
    LOOP2;
    
  }

```

https://jsfiddle.net/uhpovm9L/36/

gkjohnson replied:
This is a bit of a hack but is more along the lines of what I was thinking (https://jsfiddle.net/873rkve1/2/):

The token `UNROLLED_LOOP_INDEX` is indiscriminately replaced with the loop index when unrolling a loop so using the token in a `#define` identifier results in a new indexed identifier for each loop. For example in this case `TEST_ARR_UNROLLED_LOOP_INDEX` is unpacked into `TEST_ARR_0`, `TEST_ARR_1`, and `TEST_ARR_2` for each loop iteration respectively. This effectively lets you access array values in preprocessor macros.

I'm not sure what the project will think of this, though, but it should be significantly cleaner and more maintainable than the sorted list approach being used now imo.

```js
defines: {
  TEST_ARR_0: 2,
  TEST_ARR_1: - 1,
  TEST_ARR_2: 0,
},
  
fragmentShader: /* glsl */`
  
  uniform sampler2D textures[ 3 ];
  
  void main() {
  
	  gl_FragColor = vec4( 0.0 );
	  gl_FragColor.a = 1.0;
  
	  #pragma unroll_loop_start
	  for ( int i = 0; i < 3; i ++ ) {
  
		  #if TEST_ARR_UNROLLED_LOOP_INDEX != - 1

		  gl_FragColor += texture( textures[ TEST_ARR_UNROLLED_LOOP_INDEX ], vec2( 0.5 ) );

		  #endif
  
	  }
	  #pragma unroll_loop_end
	  
  }
`
```

unrolls into:

```glsl
void main() {

    gl_FragColor = vec4( 0.0 );
    gl_FragColor.a = 1.0;
    
    #if TEST_ARR_0 != - 1
    
    gl_FragColor += texture( textures[ TEST_ARR_0 ], vec2( 0.5 ) );
    
    #endif
    
    #if TEST_ARR_1 != - 1
    
    gl_FragColor += texture( textures[ TEST_ARR_1 ], vec2( 0.5 ) );
    
    #endif
    
    #if TEST_ARR_2 != - 1
    
    gl_FragColor += texture( textures[ TEST_ARR_2 ], vec2( 0.5 ) );
    
    #endif

}
```

richardassar replied:
This is much better, thanks.

sunag commented:
I just think it would be easier to merge if we could do it based on Nodes for WebGL as well, that seems closer to where things are going, in which case we should start supporting `LightNode` in WebGL as well. Things like atlas/caches can be done using a Node system in more intuitive ways.

richardassar commented:
I wasn't aware of any merge difficulties, unless dev has diverged a lot since I created this PR. 

Does it make sense to block this feature pending a node-based implementation? 

The changes to existing code are relatively minimal for a feature that some might find useful, unless all new features are to be node-based this could be refactored as a separate task.

mrdoob commented:
@sunag 

> I just think it would be easier to merge if we could do it based on Nodes for WebGL as well, that seems closer to where things are going, in which case we should start supporting `LightNode` in WebGL as well.

Agreed. Sounds like it'll be better to add this to the Nodes code instead.

richardassar commented:
@mrdoob @sunag Thanks, I'll adapt this into a Nodes implementation soon when I have some time.

Wagyx commented:
Hello there,
I have noticed a few issues with how the IES lights are implemented right now:
- in the IESLoader, the intensity is accounted for twice
- in the IESLoader, the symmetries are not accounted for correctly
- in the IESLoader, only type C is supported
- one of the default IES files in the webgl_lights_spotlight.html example does not follow the specifications of the IES format
- only the first column is considered, making all the spot lights symmetric
- why not use the map attributes to store the complete data and adress it properly using the inclination and azimuth of the ray direction ?

I could not figure out how to make the proper modifications to the three.js project so I have implemented a ray marcher with a better IESLoader : https://asliceofcuriosity.fr/assets/ray-marching-ies/webgl_raymarching_ies.html
I have dealt with the symmetries but not with types A and B yet.

I have made a github project with the source code:
https://github.com/Wagyx/ray-marching-ies

Mugen87 commented:
As mentioned in previous comments, it's best to reimplement this feature with the nodes system and with a fresh PR. This PR can also honor the feedback from @Wagyx.

gkjohnson commented:
> it's best to reimplement this feature with the nodes system for WebGPURenderer and with a fresh PR

I believe it's already been addd in #25238. @Wagyx it would be worth checking out that implementation and making a PR if you see issues.
